zerver/lib/narrow.py

```diff
@@ -947,37 +972,62 @@ def update_narrow_terms_containing_with_operator(
-        if channel is not None:
-            channel_id = channel.id
-    except Stream.DoesNotExist:
-        pass
 
     # Channel-level muting only applies when looking at views that
     # include multiple channels, since we do want users to be able to
     # browser messages within a muted channel.
-    if channel_id is None:
+    if channel_data is None:
+        channel_id = None
+
         rows = Subscription.objects.filter(
             user_profile=user_profile,
             active=True,
             is_muted=True,
             recipient__type=Recipient.STREAM,
         ).values("recipient_id")
         muted_recipient_ids = [row["recipient_id"] for row in rows]
+
+        # Only add the condition if we have muted channels to simplify/avoid warnings.
         if len(muted_recipient_ids) > 0:
-            # Only add the condition if we have muted channels to simplify/avoid warnings.
-            condition = not_(column("recipient_id", Integer).in_(muted_recipient_ids))
-            conditions.append(condition)
+            # Include unmuted and followed topics from muted channels
+            unmuted_followed_conditions = include_unmutes_and_followed_topics(user_profile)
+            if unmuted_followed_conditions is not None:
+                or_condition = or_(
+                    not_(column("recipient_id", Integer).in_(muted_recipient_ids)),
+                    unmuted_followed_conditions,
+                )
+                conditions.append(or_condition)
+            else:
+                condition = not_(column("recipient_id", Integer).in_(muted_recipient_ids))
+                conditions.append(condition)
+        else:
+            # If there are no muted channels, we should include messages from all channels.
+            conditions.append(true())
+        conditions = exclude_topic_mutes(conditions, user_profile, channel_id)
 
-    conditions = exclude_topic_mutes(conditions, user_profile, channel_id)
+    else:
+        channel_id = channel_data["stream"].id
+        channel_subscription = channel_data["sub"]
+
+        # If we're in a channel narrow, we need to take into consideration if the channel
+        # is muted.
+        if channel_subscription.is_muted:
+            # If channel is muted, default to excluding messages and include only unmuted topics
+            # and followed topics.
+            unmuted_followed_conditions = include_unmutes_and_followed_topics(
+                user_profile, channel_id
+            )
+            if unmuted_followed_conditions is not None:
+                conditions.append(unmuted_followed_conditions)
+            else:
+                # If there are no unmuted or followed topics, we should not include any messages.
+                conditions.append(false())
+        else:
+            # If channel is not muted, default to including messages and excluding muted topics.
+            conditions = exclude_topic_mutes(conditions, user_profile, channel_id)
```

Comment:
So this change, to include unmuted/followed topics, we'll want to carefully analyze the performance characteristics of; I could imagine it being unduly expensive. That is part of why I'd like that to be in its own commit, maybe a new final commit: It's a lot more risky than the rest of this PR.

Reply:
I've added the `explain analyze` for the entire queries to the description, but the fetching of unmuted/followed topics is done using the Django ORM just like it was being done earlier, I'm not sure what impact this has on the overall query.